src/OrchardCore/OrchardCore.BackgroundTasks/BackgroundHostedService.cs

```diff
@@ -0,0 +1,216 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using NCrontab;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Environment.Shell.Descriptor.Models;
+using OrchardCore.Environment.Shell.Models;
+using OrchardCore.Hosting.ShellBuilders;
+
+namespace OrchardCore.BackgroundTasks
+{
+    public class BackgroundHostedService : BackgroundService, IShellDescriptorManagerEventHandler
+    {
+        private static TimeSpan PollingTime = TimeSpan.FromMinutes(1);
+        private static TimeSpan MinIdleTime = TimeSpan.FromSeconds(10);
+
+        private readonly ConcurrentDictionary<string, Scheduler> _schedulers = new ConcurrentDictionary<string, Scheduler>();
+
+        private readonly IShellHost _shellHost;
+        private readonly IHttpContextAccessor _httpContextAccessor;
+
+        public BackgroundHostedService(
+            IShellHost shellHost,
+            IHttpContextAccessor httpContextAccessor,
+            ILogger<BackgroundHostedService> logger)
+        {
+            _shellHost = shellHost;
+            _httpContextAccessor = httpContextAccessor;
+            Logger = logger;
+        }
+
+        public ILogger Logger { get; set; }
+
+        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            stoppingToken.Register(() => Logger.LogDebug($"BackgroundHostedService is stopping."));
+
+            var startUtc = DateTime.UtcNow;
+
+            while (!stoppingToken.IsCancellationRequested)
+            {
+                var pollingDelay = Task.Delay(PollingTime, stoppingToken);
+
+                var shells = GetRunningShells();
+
+                Parallel.ForEach(shells, new ParallelOptions { MaxDegreeOfParallelism = 8 }, async shell =>
```

Comment:
Unfortunately, this won't do what you expect: this method is not async-aware and the delegate you pass will be executed as an async void method, which means this will method will return way before the tasks are actually complete.

Instead, you may want to directly use the underlying partitioning APIs. E.g:

```csharp
public static Task ForEachAsync<T>(this IEnumerable<T> source, Func<T, Task> body) 
{
    return Task.WhenAll(
        from partition in Partitioner.Create(source).GetPartitions(8)
        select Task.Run(async delegate
        {
            using (partition)
            {
                while (partition.MoveNext())
                {
                    await body(partition.Current);
                }
            }
        }));
}
```

```csharp
shells.ForEachAsync(async shell => { ... });
```

Reply:
@PinpointTownes thanks, just tried, you're right. I didn't see it while debugging because in the main loop there is an await on a `Task.Delay()`.

Then i tried your suggestion and it works. What do you think about the number of partitions?

 Thanks again 👍

Reply:
My pleasure :+1: